src/main/resources/com/puppycrawl/tools/checkstyle/grammar/java.g

```diff
@@ -1591,6 +1599,7 @@ newArrayDeclarator
                 warnWhenFollowAmbig = false;
             }
         :
+            (annotation)?
```

Comment:
reference: [`ArrayCreationExpression`](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-ArrayCreationExpression) which allows an optional annotation in
```
DimExpr:
  {Annotation} [ Expression ]
```

Reply:
@pbludov re:  `new String @B [3];` for [NoWhitespaceAfter](https://checkstyle.org/config_whitespace.html#NoWhitespaceAfter), with the grammar here, this parses as
```
--LITERAL_NEW -> new
   |--IDENT -> String
   `--ARRAY_DECLARATOR -> [
       |--ANNOTATION -> ANNOTATION
       |   |--AT -> @
       |   `--IDENT -> B
       |--EXPR -> EXPR
       |   `--NUM_INT -> 3
       `--RBRACK -> ]
```

Does that look like what we want the AST to look like for these expressions?

If so, maybe we need to update the logic in the check?

So in check documentation, the `final char @NotNull [] param` example parses as
```
--PARAMETER_DEF -> PARAMETER_DEF
   |--MODIFIERS -> MODIFIERS
   |   `--FINAL -> final
   |--TYPE -> TYPE
   |   `--ARRAY_DECLARATOR -> [
   |       |--LITERAL_CHAR -> char
   |       |--ANNOTATIONS -> ANNOTATIONS
   |       |   `--ANNOTATION -> ANNOTATION
   |       |       |--AT ->
   |       |       `--IDENT -> NotNull
   |       `--RBRACK -> ]
   `--IDENT -> param
```
and here in the code https://github.com/checkstyle/checkstyle/blob/55e3f7553ffc6fe4c59505273bf6209054d325b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/NoWhitespaceAfterCheck.java#L170-L174 `whitespaceFollowedAst` is `LITERAL_CHAR -> char` whose next sibling is `ANNOTATIONS`

But for `new String @B [3];`, we have `whitespaceFollowedAst` as `IDENT -> String` whose next sibling is `ARRAY_DECLARATOR` which doesn't get filtered out.

Reply:
>    `--ARRAY_DECLARATOR -> [`
>    `    |--ANNOTATION -> ANNOTATION`

It looks like there is a missing `ANNOTATIONS` node.

Reply:
@esilkensen is it possible to change the grammar to get the tree
```
   `--ARRAY_DECLARATOR -> [
       |--ANNOTATIONS -> ANNOTATIONS
          |--ANNOTATION -> ANNOTATION
              |--...
          |--ANNOTATION -> ANNOTATION
              |--...
```
to unify with all other cases for annotations?

Reply:
Yes, I will make those changes now

Reply:
Thanks @pbludov - initially I was reading `{Annotation}` in the spec as "optional," but now see that it means "zero or more" - I updated my changes to the grammar so they follow the spec and other cases for annotations in the grammar. I can update the regression reports too with these latest changes.